File: java/Samples/Biometrics/VB/SimpleFacesSampleVB/CaptureIcaoCompliantImage.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.IO
Imports System.Windows.Forms
Imports Neurotec.Biometrics
Imports Neurotec.Biometrics.Client
Imports Neurotec.Devices
Imports Neurotec.Images

Partial Public Class CaptureIcaoCompliantImage
	Inherits UserControl
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
	End Sub

#End Region

#Region "Private fields"

	Private _biometricClient As NBiometricClient
	Private _subject As NSubject
	Private _face As NFace
	Private _segmentedFace As NFace
	Private _deviceManager As NDeviceManager

#End Region

#Region "Public properties"

	Public Property BiometricClient() As NBiometricClient
		Get
			Return _biometricClient
		End Get
		Set(ByVal value As NBiometricClient)
			_biometricClient = value
		End Set
	End Property

#End Region

#Region "Private methods"

	Private Sub UpdateCameraList()
		cbCameras.BeginUpdate()
		Try
			cbCameras.Items.Clear()
			For Each device As NDevice In _deviceManager.Devices
				cbCameras.Items.Add(device)
			Next device

			If _biometricClient.FaceCaptureDevice Is Nothing AndAlso cbCameras.Items.Count > 0 Then
				cbCameras.SelectedIndex = 0
				Return
			End If

			If _biometricClient.FaceCaptureDevice IsNot Nothing Then
				cbCameras.SelectedIndex = cbCameras.Items.IndexOf(_biometricClient.FaceCaptureDevice)
			End If
		Finally
			cbCameras.EndUpdate()
		End Try
	End Sub

	Private Sub EnableControls(ByVal capturing As Boolean)
		Dim hasTemplate As Boolean = (Not capturing) AndAlso _subject IsNot Nothing AndAlso _subject.Status = NBiometricStatus.Ok
		btnSaveImage.Enabled = hasTemplate
		btnSaveTemplate.Enabled = hasTemplate
		btnStart.Enabled = Not capturing
		btnRefreshList.Enabled = Not capturing
		btnStop.Enabled = capturing
		cbCameras.Enabled = Not capturing
		btnForce.Enabled = capturing
	End Sub

	Private Sub OnCapturingCompleted(ByVal r As IAsyncResult)
		If InvokeRequired Then
			BeginInvoke(New AsyncCallback(AddressOf OnCapturingCompleted), r)
		Else
			Try
				Dim task As NBiometricTask = _biometricClient.EndPerformTask(r)
				Dim status As NBiometricStatus = task.Status

				If task.Error IsNot Nothing Then
					Utils.ShowException(task.Error)
				End If
				If status = NBiometricStatus.Ok Then
					_segmentedFace = _subject.Faces(1)
					faceView.Face = _segmentedFace
					icaoWarningView.Face = _segmentedFace
				End If

				lblStatus.Text = status.ToString()
				lblStatus.ForeColor = If(status = NBiometricStatus.Ok, Color.Green, Color.Red)
				EnableControls(False)
			Catch ex As Exception
				Utils.ShowException(ex)
				lblStatus.Text = String.Empty
				EnableControls(False)
			End Try
		End If
	End Sub

#End Region

	Private Sub CaptureIcaoCompliantImageLoad(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
		If (Not DesignMode) Then
			Try
				nViewZoomSlider2.View = faceView
				lblStatus.Text = String.Empty
				_deviceManager = _biometricClient.DeviceManager
				saveImageDialog.Filter = NImages.GetSaveFileFilterString()
				UpdateCameraList()
			Catch ex As Exception
				Utils.ShowException(ex)
			End Try
		End If
	End Sub

	Private Sub BtnRefreshListClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnRefreshList.Click
		UpdateCameraList()
	End Sub

	Private Sub CbCamerasSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbCameras.SelectedIndexChanged
		_biometricClient.FaceCaptureDevice = TryCast(cbCameras.SelectedItem, NCamera)
	End Sub

	Private Sub BtnStartClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnStart.Click
		If _biometricClient.FaceCaptureDevice Is Nothing Then
			MessageBox.Show("Please select camera from the list")
			Return
		End If

		' Set face capture from stream
		_face = New NFace With {.CaptureOptions = NBiometricCaptureOptions.Stream}
		_subject = New NSubject()
		_subject.Faces.Add(_face)
		faceView.Face = _face
		IcaoWarningView.Face = _face

		_biometricClient.FacesCheckIcaoCompliance = True

		Dim task = _biometricClient.CreateTask(NBiometricOperations.Capture Or NBiometricOperations.Segment Or NBiometricOperations.CreateTemplate, _subject)
		_biometricClient.BeginPerformTask(task, AddressOf OnCapturingCompleted, Nothing)

		lblStatus.Text = String.Empty
		EnableControls(True)
	End Sub

	Private Sub BtnStopClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnStop.Click
		_biometricClient.Cancel()
	End Sub

	Private Sub BtnSaveTemplateClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnSaveTemplate.Click
		If saveTemplateDialog.ShowDialog() = DialogResult.OK Then
			File.WriteAllBytes(saveTemplateDialog.FileName, _subject.GetTemplateBuffer().ToArray())
		End If
	End Sub

	Private Sub BtnSaveImageClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnSaveImage.Click
		If saveImageDialog.ShowDialog() = DialogResult.OK Then
			_segmentedFace.Image.Save(saveImageDialog.FileName)
		End If
	End Sub

	Private Sub BtnForceClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnForce.Click
		_biometricClient.Force()
	End Sub
End Class

File: java/Samples/Biometrics/CS/SimpleFacesSampleCS/CaptureIcaoCompliantImage.cs
using System;
using System.Drawing;
using System.IO;
using System.Windows.Forms;
using Neurotec.Biometrics;
using Neurotec.Biometrics.Client;
using Neurotec.Devices;
using Neurotec.Images;

namespace Neurotec.Samples
{
	public partial class CaptureIcaoCompliantImage : UserControl
	{
		#region Public constructor

		public CaptureIcaoCompliantImage()
		{
			InitializeComponent();
		}

		#endregion

		#region Private fields

		private NBiometricClient _biometricClient;
		private NSubject _subject;
		private NFace _face;
		private NFace _segmentedFace;
		private NDeviceManager _deviceManager;

		#endregion

		#region Public properties

		public NBiometricClient BiometricClient
		{
			get { return _biometricClient; }
			set { _biometricClient = value; }
		}

		#endregion

		#region Private methods

		private void UpdateCameraList()
		{
			cbCameras.BeginUpdate();
			try
			{
				cbCameras.Items.Clear();
				foreach (NDevice device in _deviceManager.Devices)
				{
					cbCameras.Items.Add(device);
				}

				if (_biometricClient.FaceCaptureDevice == null && cbCameras.Items.Count > 0)
				{
					cbCameras.SelectedIndex = 0;
					return;
				}

				if (_biometricClient.FaceCaptureDevice != null)
				{
					cbCameras.SelectedIndex = cbCameras.Items.IndexOf(_biometricClient.FaceCaptureDevice);
				}
			}
			finally
			{
				cbCameras.EndUpdate();
			}
		}

		private void EnableControls(bool capturing)
		{
			bool hasTemplate = !capturing && _subject != null && _subject.Status == NBiometricStatus.Ok;
			btnSaveImage.Enabled = hasTemplate;
			btnSaveTemplate.Enabled = hasTemplate;
			btnStart.Enabled = !capturing;
			btnRefreshList.Enabled = !capturing;
			btnStop.Enabled = capturing;
			cbCameras.Enabled = !capturing;
			btnForce.Enabled = capturing;
		}

		private void OnCapturingCompleted(IAsyncResult r)
		{
			if (InvokeRequired)
			{
				BeginInvoke(new AsyncCallback(OnCapturingCompleted), r);
			}
			else
			{
				try
				{
					NBiometricTask task = _biometricClient.EndPerformTask(r);
					NBiometricStatus status = task.Status;

					if (task.Error != null)
						Utils.ShowException(task.Error);
					if (status == NBiometricStatus.Ok)
					{
						_segmentedFace = _subject.Faces[1];
						faceView.Face = _segmentedFace;
						icaoWarningView.Face = _segmentedFace;
					}

					lblStatus.Text = status.ToString();
					lblStatus.ForeColor = status == NBiometricStatus.Ok ? Color.Green : Color.Red;
					EnableControls(false);
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
					lblStatus.Text = string.Empty;
					EnableControls(false);
				}
			}
		}

		#endregion

		private void CaptureIcaoCompliantImageLoad(object sender, EventArgs e)
		{
			if (!DesignMode)
			{
				try
				{
					nViewZoomSlider2.View = faceView;
					lblStatus.Text = string.Empty;
					_deviceManager = _biometricClient.DeviceManager;
					saveImageDialog.Filter = NImages.GetSaveFileFilterString();
					UpdateCameraList();
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
				}
			}
		}

		private void BtnRefreshListClick(object sender, EventArgs e)
		{
			UpdateCameraList();
		}

		private void CbCamerasSelectedIndexChanged(object sender, EventArgs e)
		{
			_biometricClient.FaceCaptureDevice = cbCameras.SelectedItem as NCamera;
		}

		private void BtnStartClick(object sender, EventArgs e)
		{
			if (_biometricClient.FaceCaptureDevice == null)
			{
				MessageBox.Show(@"Please select camera from the list");
				return;
			}

			// Set face capture from stream
			_face = new NFace { CaptureOptions = NBiometricCaptureOptions.Stream };
			_subject = new NSubject();
			_subject.Faces.Add(_face);
			faceView.Face = _face;
			icaoWarningView.Face = _face;

			_biometricClient.FacesCheckIcaoCompliance = true;

			var task = _biometricClient.CreateTask(NBiometricOperations.Capture | NBiometricOperations.Segment | NBiometricOperations.CreateTemplate, _subject);
			_biometricClient.BeginPerformTask(task, OnCapturingCompleted, null);

			lblStatus.Text = string.Empty;
			EnableControls(true);
		}

		private void BtnStopClick(object sender, EventArgs e)
		{
			_biometricClient.Cancel();
		}

		private void BtnSaveTemplateClick(object sender, EventArgs e)
		{
			if (saveTemplateDialog.ShowDialog() == DialogResult.OK)
			{
				File.WriteAllBytes(saveTemplateDialog.FileName, _subject.GetTemplateBuffer().ToArray());
			}
		}

		private void BtnSaveImageClick(object sender, EventArgs e)
		{
			if (saveImageDialog.ShowDialog() == DialogResult.OK)
			{
				_segmentedFace.Image.Save(saveImageDialog.FileName);
			}
		}

		private void BtnForceClick(object sender, EventArgs e)
		{
			_biometricClient.Force();
		}
	}
}
